### PR TITLE
deduplicate predicates optimistically

### PR DESCRIPTION
**Pull request by richardstartin**
Deduplicating predicates frequently shows up in profiles for a range of queries because the hash code of a `Predicate` is both polymorphic and relatively slow to compute. 

Screenshot 2022-05-12 at 12 35 17
Screenshot 2022-05-12 at 12 35 37

### CODE REVIEW

**Comment by richardstartin** [comment]
This doesn't have a huge impact but given that planning is single threaded, it could be worth it:, though it trades an improvement in the common case for what will be a regression in rare cases

```
Benchmark               (_numRows)                                                                                                                                                      (_query)  (_scenario)  Mode  Cnt  Score   Error  Units
BenchmarkQueries.query     1500000  SELECT INT_COL, SORTED_COL, SUM(RAW_INT_COL) FROM MyTable WHERE INT_COL = 0 and SORTED_COL = 1 GROUP BY INT_COL, SORTED_COL ORDER BY SORTED_COL, INT_COL ASC   EXP(0.001)  avgt    5  0.667 ± 0.141  ms/op
BenchmarkQueries.query     1500000  SELECT INT_COL, SORTED_COL, SUM(RAW_INT_COL) FROM MyTable WHERE INT_COL = 0 and SORTED_COL = 1 GROUP BY INT_COL, SORTED_COL ORDER BY SORTED_COL, INT_COL ASC     EXP(0.5)  avgt    5  0.500 ± 0.255  ms/op
BenchmarkQueries.query     1500000  SELECT INT_COL, SORTED_COL, SUM(RAW_INT_COL) FROM MyTable WHERE INT_COL = 0 and SORTED_COL = 1 GROUP BY INT_COL, SORTED_COL ORDER BY SORTED_COL, INT_COL ASC   EXP(0.999)  avgt    5  0.828 ± 0.117  ms/op
```

```
Benchmark               (_numRows)                                                                                                                                                      (_query)  (_scenario)  Mode  Cnt  Score   Error  Units
BenchmarkQueries.query     1500000  SELECT INT_COL, SORTED_COL, SUM(RAW_INT_COL) FROM MyTable WHERE INT_COL = 0 and SORTED_COL = 1 GROUP BY INT_COL, SORTED_COL ORDER BY SORTED_COL, INT_COL ASC   EXP(0.001)  avgt    5  0.435 ± 0.116  ms/op
BenchmarkQueries.query     1500000  SELECT INT_COL, SORTED_COL, SUM(RAW_INT_COL) FROM MyTable WHERE INT_COL = 0 and SORTED_COL = 1 GROUP BY INT_COL, SORTED_COL ORDER BY SORTED_COL, INT_COL ASC     EXP(0.5)  avgt    5  0.458 ± 0.238  ms/op
BenchmarkQueries.query     1500000  SELECT INT_COL, SORTED_COL, SUM(RAW_INT_COL) FROM MyTable WHERE INT_COL = 0 and SORTED_COL = 1 GROUP BY INT_COL, SORTED_COL ORDER BY SORTED_COL, INT_COL ASC   EXP(0.999)  avgt    5  0.598 ± 0.111  ms/op
```

File: pinot-core/src/main/java/org/apache/pinot/core/startree/StarTreeUtils.java [hunk 1]
```diff
@@ -94,13 +94,13 @@ public static AggregationFunctionColumnPair[] extractAggregationFunctionPairs(
    */
   @Nullable
   public static Map<String, List<CompositePredicateEvaluator>> extractPredicateEvaluatorsMap(IndexSegment indexSegment,
-      @Nullable FilterContext filter, Map<Predicate, PredicateEvaluator> predicateEvaluatorMap) {
+      @Nullable FilterContext filter, List<Pair<Predicate, PredicateEvaluator>> predicateEvaluatorMap) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/startree/StarTreeUtils.java:97]:
```suggestion
      @Nullable FilterContext filter, List<Pair<Predicate, PredicateEvaluator>> predicateEvaluators) {
```

File: pinot-core/src/main/java/org/apache/pinot/core/startree/StarTreeUtils.java [hunk 2]
```diff
@@ -187,7 +187,7 @@ public static boolean isFitForStarTree(StarTreeV2Metadata starTreeV2Metadata,
    */
   @Nullable
   private static Pair<String, List<PredicateEvaluator>> isOrClauseValidForStarTree(IndexSegment indexSegment,
-      FilterContext filter, Map<Predicate, PredicateEvaluator> predicateEvaluatorMap) {
+      FilterContext filter, List<Pair<Predicate, PredicateEvaluator>> predicateEvaluatorMap) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/startree/StarTreeUtils.java:190]:
```suggestion
      FilterContext filter, List<Pair<Predicate, PredicateEvaluator>> predicateEvaluators) {
```

File: pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java [hunk 3]
```diff
@@ -68,7 +67,7 @@ public class FilterPlanNode implements PlanNode {
   private final FilterContext _filter;
 
   // Cache the predicate evaluators
-  private final Map<Predicate, PredicateEvaluator> _predicateEvaluatorMap = new HashMap<>();
+  private final List<Pair<Predicate, PredicateEvaluator>> _predicateEvaluatorMap = new ArrayList<>();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java:70]:
```suggestion
  private final List<Pair<Predicate, PredicateEvaluator>> _predicateEvaluators = new ArrayList<>();
```

File: pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java [hunk 4]
```diff
@@ -108,7 +107,7 @@ public BaseFilterOperator run() {
   /**
    * Returns a map from predicates to their evaluators.
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/plan/FilterPlanNode.java:108]: Update the comment

File: pinot-core/src/main/java/org/apache/pinot/core/startree/StarTreeUtils.java [hunk 5]
```diff
@@ -258,7 +258,7 @@ private static boolean extractOrClausePredicates(FilterContext filter, List<Pred
    */
   @Nullable
   private static PredicateEvaluator getPredicateEvaluator(IndexSegment indexSegment, Predicate predicate,
-      Map<Predicate, PredicateEvaluator> predicateEvaluatorMap) {
+      List<Pair<Predicate, PredicateEvaluator>> predicatesAndEvaluators) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/startree/StarTreeUtils.java:261]: (nit) for consistency
```suggestion
      List<Pair<Predicate, PredicateEvaluator>> predicateEvaluators) {
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM. Only minor comments about naming

**Comment by richardstartin** [comment]
@Jackie-Jiang lots of comments, I changed the word "Map" to "Mapping" throughout.

**Pull request merged**